### PR TITLE
feat(systemtags): add color support

### LINKED ISSUE #6778: colored tags
As folders get more and more outdated today and tags are the new folders. I also see the future of tags in nextcloud.

While using a lot of tags it get more difficult to distinguish between them. A color per tag can make this easier.

### PR DESCRIPTION
**Pull request by skjnldsv**
fix https://github.com/nextcloud/server/issues/6778
WIP

## In progress
- [x] Add color backend support
- [x] Add bulk tagging UI color support
- [x] Add inline tags color support

## Screenshot
image

### CODE REVIEW

**Comment by skjnldsv** [comment]
Amazing... :facepalm: 
image

**Review by jancborchardt — APPROVED** [review]
Looks very very nice! I have no feedback. @marcoambrosini?

**Review by jancborchardt — DISMISSED** [review]
Sorry actually I do have 2 pieces of feedback :D 
- [ ] The popover seems to be very slightly misaligned to the left:
   Screenshot From 2024-11-20 14-10-17
- [x] There is a stray triangle here: 
   image

**Comment by marcoambrosini** [comment]
Yep, just those things, which both seem related to the popover positioning. I'll wait for review ping

**Comment by skjnldsv** [comment]
> * There is a stray triangle here

Lol, it's the scrollbar behind :grin:

**Comment by skjnldsv** [comment]
@jancborchardt There is also the inline tag display missing on this PR.
But I read many of your comments on the topic from [here](https://github.com/nextcloud/server/issues/6778) and [here](https://github.com/nextcloud/server/issues/6424

I'll keep them like they are right now, no background and color the text+border only, following the WCAG contrast of course.

Quick example (⚠️  contrast not calculated)
image

File: lib/private/SystemTag/SystemTagManager.php [hunk 1]
```diff
@@ -218,19 +204,23 @@ public function updateTag(
...
 			->set('visibility', $query->createParameter('visibility'))
 			->set('editable', $query->createParameter('editable'))
+			->set('color', $query->createParameter('color'))
 			->where($query->expr()->eq('id', $query->createParameter('tagid')))
 			->setParameter('name', $truncatedNewName)
 			->setParameter('visibility', $userVisible ? 1 : 0)
 			->setParameter('editable', $userAssignable ? 1 : 0)
-			->setParameter('tagid', $tagId);
+			->setParameter('tagid', $tagId)
+			->setParameter('color', $color);
```
Comment (marcelklehr) [inline lib/private/SystemTag/SystemTagManager.php:223]: Should we validate the color value in the backend? Currently users of the DAV endpoint and the OCP API can just insert anything, I think.

Comment (artonge) [reply lib/private/SystemTag/SystemTagManager.php:223]: Other option would be to change it during the display to be accessible

**Review by artonge — APPROVED** [review]

File: lib/private/SystemTag/SystemTagManager.php (hunk 1 above)
Comment (marcelklehr) [reply lib/private/SystemTag/SystemTagManager.php:223]: So we have to validate on every client that wants to display colors?

Comment (marcelklehr) [reply lib/private/SystemTag/SystemTagManager.php:223]: This query will throw an SQL exception when you pass 7 characters for example

Comment (skjnldsv) [reply lib/private/SystemTag/SystemTagManager.php:223]: let me add a quick check :+1:

**Review by marcelklehr — APPROVED** [review]

**Comment by disaster123** [comment]
Thanks a lot for this. What is the planend release Version?

**Comment by marcelklehr** [comment]
@disaster123 Nextcloud 31 Hub 10

**Pull request merged**